Modernize use of UTF-8 in Windows code

hebasto opened the pull request:
The main goal is to remove [deprecated](https://github.com/bitcoin/bitcoin/issues/32361) code (removed in C++26).

This PR employs Microsoft's modern [approach](https://learn.microsoft.com/en-us/windows/apps/design/globalizing/use-utf8-code-page) to handling UTF-8:
> Until recently, Windows has emphasized "Unicode" -W variants over -A APIs. However, recent releases have used the ANSI code page and -A APIs as a means to introduce UTF-8 support to apps. If the ANSI code page is configured for UTF-8, then -A APIs typically operate in UTF-8. This model has the benefit of supporting existing code built with -A APIs without any code changes.

TODO:
- [x] Handle application manifests properly when building with MSVC.
- [x] Bump the minimum supported Windows version to 1903 (May 2019 Update).
- [x] Remove all remaining use cases of the deprecated `std:wstring_convert`.
    - The instance in `subprocess.h` will be addressed in a follow-up PR, as additional tests are likely needed.
    - The usage in `common/system.cpp` is handled in https://github.com/bitcoin/bitcoin/pull/32566.

Resolves partially https://github.com/bitcoin/bitcoin/issues/32361.

hebasto commented:
Rebased on https://github.com/bitcoin/bitcoin/pull/32383.

laanwj commented on src/common/args.cpp:
```diff
@@ -867,27 +867,15 @@ void ArgsManager::LogArgs() const
+WindowsScopedCodePage::~WindowsScopedCodePage()
 {
-    return std::make_pair(argc, argv);
+    SetConsoleOutputCP(m_original_code_page);
```
Is this necessary with the manifest?

Also we already have
```
    SetConsoleCP(CP_UTF8);
    SetConsoleOutputCP(CP_UTF8);
```
in `void SetupEnvironment()`.

hebasto replied:
> Is this necessary with the manifest?

The manifest is responsible for the active code page, which differs from console input and output code pages. It would be easy to compare outputs of `GetACP()`, `GetConsoleCP()` and `GetConsoleOutputCP()`.

> Also we already have
> 
> ```
>     SetConsoleCP(CP_UTF8);
>     SetConsoleOutputCP(CP_UTF8);
> ```
> 
> in `void SetupEnvironment()`.

Thanks! I overlooked those calls. However, with the manifest, setting only the console output code page is sufficient.

>  It's not necessary to scope this as we want it for the entire run of the application.

The scoped implementation is necessary to restore the console code pages to their state prior to the application's execution.

laanwj replied:
> The manifest is responsible for the active code page, which differs from console input and output code pages. It would be easy to compare outputs of GetACP(), GetConsoleCP() and GetConsoleOutputCP().

Thanks.

> Thanks! I overlooked those calls. However, with the manifest, setting only the console output code page is sufficient.

OK.

> The scoped implementation is necessary to restore the console code pages to their state prior to the application's execution.

We haven't been doing this ever. I think it was nice to simply roll this into `SetupEnvironment`. But okay.

laanwj replied:
If we really need to, what if we make `SetupEnvironment` scoped instead? It's a more general solution, it doesn't belong in "common/args", and would work also if there's future environment modifications that need reverting before exit.

laanwj commented:
After this we should revert our custom [leveldb unicode patch](f8ae182c1e5176d12e816fb2217ae33a5472fdd7). This would make the `env_windows` backend go back to using the `A` functions, which already take UTF-8 as-is.

hebasto commented:
Rebased on https://github.com/bitcoin/bitcoin/pull/32396.

hebasto committed:
Remove no longer necessary `WinCmdLineArgs` class

This change removes one use case of `std::wstring_convert`, which is
deprecated in C++17 and removed in C++26. Other uses remain for now.

hebasto committed:
Switch to ANSI Windows API in `fsbridge::fopen()` function

hebasto commented:
Rebased on https://github.com/bitcoin/bitcoin/pull/32537.

hebasto committed:
Switch to ANSI Windows API in `Win32ErrorString()` function

hebasto committed:
Set minimum supported Windows version to 1903 (May 2019 Update)

This version is the minimum required to support the UTF-8 code page
(CP_UTF8).

fanquake commented on doc/release-notes-empty-template.md:
```diff
@@ -35,8 +35,8 @@ wallet versions of Bitcoin Core are generally supported.
-Bitcoin Core is supported and tested on operating systems using the
-Linux Kernel 3.17+, macOS 13+, and Windows 10+. Bitcoin
+Bitcoin Core is supported and tested on the following operating systems or newer:
+Linux Kernel 3.17, macOS 13, and Windows 10 (version 1903). Bitcoin
```
Is anything enforcing this, or is the assumption that any Windows 10 system is this at least version? What happens if it isn't?

hebasto replied:
> Is anything enforcing this, or is the assumption that any Windows 10 system is at least this version?

The latter.

> What happens if it isn't?

My guess is that the application won't properly handle any symbols outside its default code page.

laanwj replied:
Can we maybe check the codepage early at startup (e.g. `SetupEnvironment()`) and fail if it isn't correct? This seems more thorough than any version check.

hebasto replied:
> Can we maybe check the codepage early at startup (e.g. `SetupEnvironment()`) and fail if it isn't correct? This seems more thorough than any version check.

Thanks! Done.

hebasto commented:
Changes overlapping with https://github.com/bitcoin/bitcoin/pull/32566 have been dropped.

fanquake commented on cmake/windows-app.manifest.in:
```diff
@@ -1,10 +1,15 @@
       name="org.bitcoincore.${target}"
       version="${CLIENT_VERSION_MAJOR}.${CLIENT_VERSION_MINOR}.${CLIENT_VERSION_BUILD}.0"
   />
+  <asmv3:application>
```
In 1e826bf91932e2d76969ae40a0c0ef85a990cf2f: Can the commit be updated to better explain why this is needed. Looking at mingw-w64, `GetACP` has returned `CP_UTF8` since it was introduced. I'm guessing the same is happening with MSVC (although that's less relevant given it doesn't effect releases)? Given that, and the fact that according to https://github.com/bitcoin/bitcoin/pull/32380/files#r2096044476, we are just going to assume that a user is running a new enough version of Windows, and throw an assertion if the codepage isn't `CP_UTF8`, what is the app-manifest actually doing? Given the version setting from #32537 also didn't work, it's not clear why any of this is needed.

hebasto replied:
> In [1e826bf](https://github.com/bitcoin/bitcoin/commit/1e826bf91932e2d76969ae40a0c0ef85a990cf2f): Can the commit be updated to better explain why this is needed. Looking at mingw-w64, `GetACP` has returned `CP_UTF8` since it was introduced. I'm guessing the same is happening with MSVC (although that's less relevant given it doesn't effect releases)?

`GetACP` being compiled on Windows returns 1252, which is not the UT8 code page. And, basically, it would depend on the OS language settings.

laanwj replied:
fwiw: `GetACP` is a windows system function implemented in `kernel32.dll`, defined in the `winnls.h` header. There shouldn't be a mingw-specific definition.

`CP_UTF8` is also defined in that header and has the following value:
```c++
#define CP_UTF8 65001
```

`1252` is codepage [windows-1252](https://en.wikipedia.org/wiki/Windows-1252), a legacy encoding.

hebasto commented:
Rebased to refresh the CI.

hodlinator reviewed (commented):
Reviewed aa830373e87325f08b647d41ac3ab5f5196e46c8

Built a Guix build and tested on Windows. Verified all installed binaries worked. Also verified natively built bitcoind worked.

<details><summary>Checked embedded resources using mt</summary>

```
C:\Program Files\Bitcoin\daemon>mt -inputresource:bitcoind.exe;#1 -out:con
Microsoft (R) Manifest Tool
Copyright (c) Microsoft Corporation.
All rights reserved.
﻿<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0" xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
  <assemblyIdentity type="win32" name="org.bitcoincore.bitcoind" version="30.99.0.0"></assemblyIdentity>
  <asmv3:application>
    <asmv3:windowsSettings xmlns="http://schemas.microsoft.com/SMI/2019/WindowsSettings">
      <activeCodePage>UTF-8</activeCodePage>
    </asmv3:windowsSettings>
  </asmv3:application>
  <trustInfo xmlns="urn:schemas-microsoft-com:asm.v3">
    <security>
      <requestedPrivileges>
        <requestedExecutionLevel level="asInvoker" uiAccess="false"></requestedExecutionLevel>
      </requestedPrivileges>
    </security>
  </trustInfo>
</assembly>

C:\Program Files\Bitcoin>mt -inputresource:C:\Users\hodlinator\bitcoin\build\bin\Release\fuzz.exe;#1 -out:con
Microsoft (R) Manifest Tool
Copyright (c) Microsoft Corporation.
All rights reserved.
﻿<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0" xmlns:asmv3="urn:schemas-microsoft-com:asm.v3">
  <assemblyIdentity type="win32" name="org.bitcoincore.fuzz" version="30.99.0.0"></assemblyIdentity>
  <asmv3:application>
    <asmv3:windowsSettings xmlns="http://schemas.microsoft.com/SMI/2019/WindowsSettings">
      <activeCodePage>UTF-8</activeCodePage>
    </asmv3:windowsSettings>
  </asmv3:application>
  <trustInfo xmlns="urn:schemas-microsoft-com:asm.v3">
    <security>
      <requestedPrivileges>
        <requestedExecutionLevel level="asInvoker" uiAccess="false"></requestedExecutionLevel>
      </requestedPrivileges>
    </security>
  </trustInfo>
</assembly>
```

</details>

Checked out master (5aec516b2ce327227ad478ef798828ca3a3db81a) and cherry-picked `system.cpp` change to add (+ added `fprintf`):
```
    fprintf(stderr, "GetACP(): %d\n", GetACP());
    assert(GetACP() == CP_UTF8);
```
Ran natively built bitcoind:
```
$ ./build/bin/Release/bitcoind
GetACP(): 1252
Assertion failed: GetACP() == CP_UTF8, file C:\Users\hodlinator\bitcoin\src\common\system.cpp, line 87
```

hebasto committed:
cmake: Set process code page to UTF-8 on Windows

Additionally, this change adds app manifests to targets that were
previously missing them.

hebasto commented:
The [feedback](https://github.com/bitcoin/bitcoin/pull/32380#discussion_r2367130713) from @hodlinator has been addressed.

hodlinator reviewed (approved):
ACK b74a92cdb2c4d9871865a67162f47f97fb7e642c

Latest push adds missing/previously avoided call to `SetupEnvironment()` (discovered in https://github.com/bitcoin/bitcoin/pull/32380#discussion_r2367130713), this entails also adding a dependency from bitcoin.exe on bitcoin_common.

Tested Windows-native build, running `bitcoin node`.

hebasto commented:
Rebased to resolve a conflict with the merged https://github.com/bitcoin/bitcoin/pull/33229.

hodlinator reviewed (approved):
re-ACK 63f3ea2b041db56487bfb92813e32c0bfef6faa1

Confirmed to resolve conflict with #33229.

hodlinator reviewed (approved):
re-ACK 53e4951a5b5b9d166d278db4240513d09b447f58

davidgumberg reviewed (approved):
untested crACK https://github.com/bitcoin/bitcoin/commit/53e4951a5b5b9d166d278db4240513d09b447f58

Don't have access to a Windows device at the moment but I'll try and test this at some point, would there be a reasonable unit/functional test that `fsbridge::fopen` and argv can still handle non-ascii stuff on Windows? Might be nice to have, but not blocking by an means.

davidgumberg commented:
> I believe it is handled by unit and functional tests:
> 
> ```
> $ git grep ₿_🏃
> src/test/dbwrapper_tests.cpp:    fs::path ph = m_args.GetDataDirBase() / "test_runner_₿_🏃_20191128_104644";
> [...]
> ```

Ahh, I missed that, I thought 🏃 was just for show, but it serves a useful purpose, thanks!
